### PR TITLE
Add dll_mutate_testcase_ptr

### PR DESCRIPTION
**Pull request by aegiryy**
This PR allows to implement a customized mutation from a DLL

### CODE REVIEW

**Comment by ifratric** [comment]
Hey, plugging in a custom mutator is really cool! But I think this should be documented a bit as it's not that trivial to write (e.g. the custom mutator must call common_fuzz_stuff and also return specific values depending on the return values of common_fuzz_stuff). An usage example in the README showing a minimal custom mutator would also be nice, to make it easier for other people to use this.

**Comment by ifratric** [comment]
Also, what is the reason for passing `queue_cur->fname` to the mutator? Shouldn't the content passed via `in_buf` and `len` be sufficient?

**Comment by aegiryy** [comment]
I think it's a very good suggestion to add an example to show how it works!  You are right, the name is not really necessary, let me remove it.

**Comment by aegiryy** [comment]
@ifratric I made the change according to your feedback.  Please take a look and let me know if there's anything I can help!  Thanks for your time!

**Comment by ifratric** [comment]
Hey, after a second look, I have some more questions (sorry if I misunderstood something):

 - You placed custom mutation before "TRIMMING " and "PERFORMANCE SCORE" phases. Is there a reason for this? Shouldn't the mutation come afterwards (before the "SIMPLE BITFLIP" phase)?

 - If dll_mutate_testcase returns nonzero, you jump to abandon_entry, that part is clear. However, what is the correct behavior if it does not bail out? IIUC currently, you'll just continue to all other phases including bitflips, havoc etc. Is this really desirable with a custom mutator? I thought the idea was to replace these built-in mutators?

**Comment by aegiryy** [comment]
I placed the custom mutator above all the built-in mutators so that the custom mutator has a chance to skip all of them (by returning non-zero).  Meanwhile, it can also choose to go through all the built-in mutations by returning zero.

In our use case, we use the custom mutator to implement a concolic execution based mutator (aka. whitebox fuzzing), but still want to leverage all the awesome existing mutators baked in AFL to maximize bug finding.

In summary, the current implementation allows the  custom mutator to bypass/leverage all built-in mutations based on its return value.

**Comment by ifratric** [comment]
Sorry for the delay!

Regarding mutator placing, I understand you'd want to give the mutator the chance to skip all other mutators, the question is mostly whether we consider "TRIMMING " and "PERFORMANCE SCORE" mutators at all. My understanding is that trimming attempts to shorten the sample while still retaining the same coverage (consider that a mutation that generated the current sample could have produced an overly long sample by appending data, but just a portion of this data is sufficient to trigger new coverage, thus you'd want to shorten it to focus on actually important bits when mutating it). And "PERFORMANCE SCORE" doesn't actually do mutations, but calculates a score that can be used by the mutations later on. So you'd potentially want to do these phases regardless of the mutation strategy?

Bypassing/leveraging existing mutations SGTM, but please also document this.

**Comment by aegiryy** [comment]
@ifratric I have moved the custom mutator to be after TRIMMING and PERFORMANCE SCORE.  I also updated README.md to make it more explicit that the custom mutator can skip the built-in mutations by returning a non-zero value (aka. bailout).  Let me know if my changes look good to you.

**Comment by ifratric** [comment]
Cool, thanks!

**Comment by andreafioraldi** [comment]
Hi,
cool that you want to update WinAFL with custom mutators.
Btw trimming is not compatible with custom mutation (e.g. you maintain a protobuf structure of the testcase, trimming will break it) and so it needs to be disabled or handled by the custom mutator.

In AFL++ we defined a very expressive custom mutator API that can be used to implement structured fuzzers like AFLSmart, https://github.com/AFLplusplus/AFLplusplus/blob/stable/docs/custom_mutators.md#custom-mutation, you can borrow them and implement the same API for WinAFL.
